runtime/major_gc.c

```diff
@@ -1231,12 +1231,100 @@ static intnat ephe_sweep (caml_domain_state* domain_state, intnat budget)
   return budget;
 }
 
+static void cycle_major_heap(caml_domain_state* domain,
+                             uintnat num_domains_in_stw)
+{
+  /* Cycle major heap */
+  // FIXME: delete caml_cycle_heap_stw and have per-domain copies of the data?
+  caml_cycle_heap_stw();
+  caml_gc_log("GC cycle %lu completed (heap cycled)",
+              (long unsigned int)caml_major_cycles_completed);
+
+  caml_major_cycles_completed++;
+  caml_gc_message(0x40, "Starting major GC cycle\n");
+
+  if (atomic_load_relaxed(&caml_verb_gc) & 0x400) {
+    struct gc_stats s;
+    intnat heap_words, not_garbage_words, swept_words;
+
+    caml_compute_gc_stats(&s);
+    heap_words = s.heap_stats.pool_words + s.heap_stats.large_words;
+    not_garbage_words = s.heap_stats.pool_live_words
+      + s.heap_stats.large_words;
+    swept_words = domain->swept_words;
+    caml_gc_log ("heap_words: %"ARCH_INTNAT_PRINTF_FORMAT"d "
+                 "not_garbage_words %"ARCH_INTNAT_PRINTF_FORMAT"d "
+                 "swept_words %"ARCH_INTNAT_PRINTF_FORMAT"d",
+                 heap_words, not_garbage_words, swept_words);
+
+    if (caml_stat_space_overhead.heap_words_last_cycle != 0) {
+      /* At the end of a major cycle, no object has colour MARKED.
+
+         [not_garbage_words] counts all objects which are UNMARKED.
+         Importantly, this includes both live objects and objects which are
+         unreachable in the current cycle (i.e, garbage). But we don't get
+         to know which objects are garbage until the end of the next cycle.
+
+         live_words@N = not_garbage_words@N - swept_words@N+1
+
+         space_overhead@N =
+         100.0 * (heap_words@N - live_words@N) / live_words@N
+      */
+      double live_words_last_cycle =
+        caml_stat_space_overhead.not_garbage_words_last_cycle - swept_words;
+      double space_overhead =
+        100.0 * (double)(caml_stat_space_overhead.heap_words_last_cycle
+                         - live_words_last_cycle) / live_words_last_cycle;
+
+      if (caml_stat_space_overhead.l == NULL ||
+          caml_stat_space_overhead.index == BUFFER_SIZE) {
+        struct buf_list_t *l =
+          (struct buf_list_t*)
+          caml_stat_alloc_noexc(sizeof(struct buf_list_t));
```

Comment:
Shouldn't the result of `caml_stat_alloc_noexc` checked against `NULL` here? (I know this is not new code but code moved from `cycle_all_domains_callback`)

Reply:
I think so. We should probably create an independent issue.